Suggest changes to SE skills section (#231)
CaptainSifff/paper_teaching-learning-RSE: competencies.md

(the "R") and a software engineer (the "SE") and, therefore, requires
competencies in both fields. RSEs typically apply their knowledge and
experience in larger teams which allows them to cultivate this hybrid nature.

> Therefore, we categorise the competencies into three categories: *technical skills*,

Author: I thought about this again, but next to "research skills" and "communication skills" I still think "technical skills" is the most appropriate among the suggestions circulated thus far. "Communication" also does not pick up anything from RSE, so that is not the pattern here.

> Therefore, we categorise the competencies into three categories: *technical skills*,

Reviewer: I agree with @annalenalamprecht here. I've recently talked about some of the major differences between SE and RSE, and things like navigating software citation and publication have a strong technical component, but not a software engineering component. (In other contexts one could frame them as being the skills that set RSE apart from SE, but calling them "research software engineering skills" would drop us deep into infinite recursion.)

> Therefore, we categorise the competencies into three categories: *technical skills*,

Reviewer 2: The more I think about it the more I agree with technical skills. So fine by me.

> Therefore, we categorise the competencies into three categories: *technical skills*,

Reviewer 3: ```suggestion
Therefore, we categorise the competencies into three categories: *softeware/technical skills*,
```
In the meantime I like the "/" construct better. On the one hand we still make it obvious that we are for software, on the other hand we are an instance of Research *T*echnical professionals.

(**Project team size**). Finally, different sets of skills are emphasised in
the different RSE specialisations (**RSE specialisations**).

> ## Technical skills

Reviewer 3: ```suggestion
## Software/Technical skills
```
Be consistent.

\newcommand{\skillsection}[1]{\hypertarget{skills-#1}{%
\subsubsection{\glsentrydesc{#1} (\glsentrytext{#1})}\label{skills-#1}}}

The technical skills required by an RSE overlap to a large extent with the common 
fundamental software engineering skills (see, e.g., @Landwehr2017), but put greater
emphasis on aspects related to achieving good scientific practice. 
For example, RSEs need to know how to make research software adhere to the 
\ac{FAIR4RS} principles, and how to achieve different levels of research 
software reusability @ChueHong2014. To reflect this, the technical skills
listed below complement competencies regarding the standard life-cycle of 
software development with RSE-specific focus skills. 

<!-- Adapting to the software life-cycle -->
\skillsection{SWLC}

The traditional software development life-cycle defines the stages that form the process of building a piece of software.

> Initial development generally involves an analytic process where needs and ideas are gathered and analysed (requirements engineering),

Reviewer 4: I had a similar feeling, but I think only because this sentence is now removed:

> Often the development is iterated.

Was this intentional? Because I don't see any conflict with the rest of the additions.

I do like using the formal terms here. I understand that these are still totally part of iterative development, just in more than one iteration.

> Initial development generally involves an analytic process where needs and ideas are gathered and analysed (requirements engineering),

Reviewer: @annalenalamprecht Will a reviewer with an RE background agree with this definition of requirements engineering?

I'm asking because a) I think it is a good definition (for the RSE context), but b) RE (at least in SWEBOK) has a systematic component ("the systematic handling of requirements"), which I think c) is seldomly found in RSE. Not because of subpar practice necessarily, but more due to the experimental and highly agile nature of much of RSEng.

I don't have strong feelings, but if you feel that this is relevant, perhaps this could be picked up in the paragraphs below, where you talk about the research software life-cycle? (Although c) is also only based on intuition with no reliable data to support. Sooo much still to do!)

followed by a formulation of a plan to fulfil them (design) that is finally turned into running code (implementation).
This is accompanied by different measures of quality control (e.g., reviews, testing), validating and verifying 
that things work as expected and that they continue to do when develoment progresses further.
Often the development cycles are executed iteratively and incrementally.
The life-cycle further includes periods of deployment, maintenance and further development (software evolution), 
as well as software retirement. 
Additionally, the research software life-cycle extends the traditional life-cycle
with \gls{software-publication}. The RSE should be aware of this life-cycle

> and be able to predict and cater to the changing needs a software project as it moves through the stages.

Reviewer 5: ```suggestion
and be able to predict and cater to the changing needs of a software project as it moves through the stages.
```

> and be able to predict and cater to the changing needs of a software project as it moves through the stages.

Reviewer 5: I don't see, how to predict changing needs beyond the most obvious ones.

Author: @mschwarzmeier What exactly is your question/point here?

Reviewer 5: I am not sure, I fully understand this sentence. Maybe an example could help?

Reviewer 2: This has just been moved from a different position in the paper is not really a new addition.
So let's just keep it as is for now.

Reviewer 3: Reminder to Flo: I think currently it's duplicated.

The technical skills required by an RSE overlap to a large extent with the common 
fundamental software engineering skills (see, e.g., @Landwehr2017), but put greater
emphasis on aspects related to achieving good scientific practice. 
For example, RSEs need to know how to make research software adhere to the 
\ac{FAIR4RS} principles, and how to achieve different levels of research 
software reusability @ChueHong2014. To reflect this, the technical skills
listed below complement competencies regarding the standard life-cycle of 
software development with RSE-specific focus skills. 

<!-- Adapting to the software life-cycle -->
\skillsection{SWLC}

The traditional software development life-cycle defines the stages that form the process of building a piece of software.
Initial development generally involves an analytic process where needs and ideas are gathered and analysed (requirements engineering),
followed by a formulation of a plan to fulfil them (design) that is finally turned into running code (implementation).
This is accompanied by different measures of quality control (e.g., reviews, testing), validating and verifying 

> that things work as expected and that they continue to do when develoment progresses further.

Reviewer 3: ```suggestion
that things work as expected and that they continue to do when develoment progresses further. Depending on the software project, this can mean a simple "Think-before-you-do", or more elaborate and formal processes.
```

Often the development cycles are executed iteratively and incrementally.
The life-cycle further includes periods of deployment, maintenance and further development (software evolution), 

> as well as software retirement. 

Reviewer 3: ```suggestion
as well as software retirement.
```

requires knowledge of using build systems to enable interoperability with other
systems.

<!-- Use repositories -->
\skillsection{SWREPOS}

> The RSE should be able to identify and use fitting public platforms (so-called software repositories or repos with version control) 

Author: not mentioning version control here is fine with me

> The RSE should be able to identify and use fitting public platforms (so-called software repositories or repos with version control) 

Reviewer: As far as I can see in current `main`, "version control" as a skill is only mentioned in passing in Appendix A. This is widely at odds with current teaching practice, where VCS is one of the early and central (perhaps *the* central) skill taught to researchers (see, e.g., The Carpentries curricula, but also Helmholtz & DLR practice). 

I see the point you are making @CaptainSifff, but think it'd be _very_ odd not to mention version control here at all!

In fact, I think that the current wording is a bit too abstract to explain the use of SWREPOS in the tables further down, where, e.g., Table 1 says "Should seamlessly interact with the repository of their project.", where *repository* to me clearly points to a version control repo! I'll make a change suggestion to that effect.

> The RSE should be able to identify and use fitting public platforms (so-called software repositories or repos with version control) 

Reviewer: ```suggestion
The RSE should be able to identify and use fitting public platforms (so-called software source code repositories or "repos" with version control) 
```

Reviewer: This and below comment because this skill (to me!) clearly refers to source code repos on remote public platforms, as opposed to local source code repos, or remote repos of another kind (e.g., open access repositories).

> The RSE should be able to identify and use fitting public platforms (so-called software repositories or repos with version control) 

Reviewer 2: And even thinking of the future of automation. I'd still want some form of versioning to be able
to point to the specific version I used to do my computations.
So I'd still have to somehow interact with it.

> The RSE should be able to identify and use fitting public platforms (so-called software repositories or repos with version control) 

Reviewer 2: Oh and who would be most likely to automate the version control aspect of research?
Us RSEs probably xD So we should still know how it works to be able to maintain the software.

> to share the artefacts they have created and invite the public to scrutinise them for public review.

Reviewer: ```suggestion
to share the code artefacts they have created and invite the public to scrutinise them in public reviews.
```

Reviewer 5: I know the skill is called SWREPOS, but I kinda liked the old formulation, where the artefacts were less defined. It is not a strong opinion though.

In our work on simulation software and models, it is fundamental for the understanding, that there are jupyter notebooks, showing some charts/.... For further development they are essential and therefore cannot be separated from the code. Of course the notebooks/created graphs also belong to the publications we are doing with that code.
In another project of my group, these graphs were actually displayed on gitlab-pages, so there was a tight coupling to VCS.

Reviewer 3: > I think I get your point, and have no strong feelings about this one particular change here (but do about the one above).
> 
> On the other hand, whether you commit notebooks, source files, config files, etc., these are all code (even Markdown files and - one could argue - ASCII/plaintext files). Similarly, the pages rendering are what you do with the code files, so are derived artifacts (and code themselves, HTML, CSS, etc.)...
> 
> Anyway, I'll let others be the judge of whether this change suggestion should be accepted.

I'm in @mschwarzmeier team here. I chose the wording with the intent of also being inclusive future developments.

Reviewer 6: I agree with @CaptainSifff and @mschwarzmeier, I prefer the less well defined formulation: code implies to me something that gets transformed. As much as it pains me, sometimes it is really useful to store outputs in a repo. In some cases (like notebooks) you don't have a choice. In any case I don't think we gain from sharpening the term, if anything we will confuse people.

<!-- Software behaviour awareness and analysis -->
\skillsection{MOD}

We define this as a certain quality of analytical thinking that enables an RSE to

> form a mental model of a piece of software in a specific environment (program comprehension).

Reviewer 3: I like that word. Although I wonder whether it is as comprehensive as the MOD skill over here. Which as @unode pointed out is actually more like the "Scientific Method".

<!-- Use repositories -->
\skillsection{SWREPOS}

> The RSE should be able to identify and use fitting public platforms (so-called software source code repositories or "repos" with version control) 

Reviewer 3: ```suggestion
The RSE should be able to identify and use fitting public platforms (so-called software repositories or "repos")
```

> to share the code artefacts they have created and invite the public to scrutinise them in public reviews.

Reviewer 3: ```suggestion
to share the artefacts they have created and invite the public to scrutinise them in public reviews.
```

| \gls{DOCBB}      | Should be able to write reusable building blocks.                                                                             | Same as junior, but the quality should set the standard for the project, while following current best practices.                      | Should know the current best practices and point their staff to the right resources.                                                                                                |
| \gls{LIBS}       | Should be able to use package distribution platforms.                                                                         | Same as junior, but should also be familiar with current best practices for building and deploying packages.                          | Should ensure that their project is available via an up-to-date and secure distribution platform.                                                                                   |

> | \gls{SWREPOS}    | Should seamlessly interact with the repository of their project.                                                              | Should be well-versed in the intricacies of a repository, and probably interact with repositories of multiple projects.               | Should promote the use of repositories and be able to convey best practices to junior and senior RSEs.                                                                              |

Reviewer 3: ```suggestion
| \gls{SWREPOS}    | Should seamlessly interact with the repository of their project.                                                              | Should be well-versed in the intricacies of a repository, and probably interact with repositories of multiple projects.               | Should promote the use of repositories and be able to convey best practices of sharing and reviewing to junior and senior RSEs.                                                                              |
```

Core modules are of course drawn from the three pillars of the RSE and can be categorised accordingly.

- Technical skills:
  - As a foundational module, we have an introduction to programming. Students learn at least two languages: a language that facilitates prototyping and data processing (e.g., \gls{Python} or \gls{R}) and a language for designing complex, performance-critical systems (e.g., \gls{C}/\gls{Cpp}). This exposes them to computers in a hands-on fashion and is the foundation for (\gls{DOCBB}, \gls{LIBS}).
  - Digital ecosystem module: Programming languages are not enough to work in a digital ecosystem, hence we require something like software craftsmanship, where tools such as the Unix shell, version control systems, build systems, documentation generators, package distribution platforms, and software discovery systems are taught to strengthen skills in (\gls{LIBS}, \gls{DOCBB}, \gls{SWREPOS}, \gls{SRU}).

>   - Software engineering module: Here we develop foundational software engineering competencies (basic knowledge and skill regarding requirements engineering, software architecture and design, implementation, quality assurance, software evolution), practicalities of carrying out a software project (e.g., use of version control and other software development support tools), again emphasizing and strengthening (\gls{DOCBB}, \gls{LIBS}) on a more abstract level.

Reviewer 3: The meaning has changed here.... The Software Development tools, we would have placed into Digital Ecosystem Module, whereas the architecture part over here, is more like the high-level view of how to grow big softwares.

Author: I guess we can remove the "practicalities ..." part here, especially as the importance of practical projects is emphasized a little later in the paper.

Core modules are of course drawn from the three pillars of the RSE and can be categorised accordingly.

- Software/Technical skills:
  - Foundational module: Here we have an introduction to programming: Emphasizing use cases over programming paradigms, students learn at least two languages: a language that facilitates prototyping and data processing (e.g., \gls{Python} or \gls{R}) and a language for designing complex, performance-critical systems (e.g., \gls{C}/\gls{Cpp}). This exposes them to computers in a hands-on fashion and is the foundation for (\gls{DOCBB}, \gls{LIBS}).
  - Computing environment module: Programming languages are not enough to work in a landscape of many interconnected software components; hence we require something like software craftsmanship, where tools such as the Unix shell, version control systems, build systems, documentation generators, package distribution platforms, and software discovery systems are taught to strengthen skills in (\gls{LIBS}, \gls{DOCBB}, \gls{SWREPOS}, \gls{SRU}).

>   - Software engineering module: Here we develop foundational software engineering competencies (basic knowledge and skill regarding requirements engineering, software architecture and design, implementation, quality assurance, software evolution), practicalities of carrying out a software project (e.g., use of version control and other software development support tools), again emphasizing and strengthening (\gls{DOCBB}, \gls{LIBS}) on a more abstract level.

Reviewer 3: ```suggestion
  - Software engineering module: Here we develop foundational software engineering competencies (basic knowledge and skill regarding requirements engineering, software architecture and design, implementation, quality assurance, software evolution), again emphasizing and strengthening (\gls{DOCBB}, \gls{LIBS}) on a more abstract level.
```
